lib/ipam/ipam.go

```diff
@@ -1383,6 +1349,69 @@ func (c ipamClient) convertBackendToIPAMConfig(cfg *model.IPAMConfig) *IPAMConfi
 	}
 }
 
+// ensureConsistentAffinity retrieves the pool and node for the given block and determines
+// if the pool still selects node. If it no longer matches, it will release the block
+// affinity for that node.
+// Returns a bool indicating if the block affinity was released.
+func (c ipamClient) ensureConsistentAffinity(ctx context.Context, b *model.AllocationBlock) (bool, error) {
```

Comment:
Potential optimization here would be to alter the signature like so:


```
func (c ipamClient) ensureConsistentAffinity(ctx context.Context, b *model.AllocationBlock) (*model.AllocationBlock, error) {
```

If we return a *model.AllocationBlock, we can return the input `b` when we haven't modified the block, and return the re-queried block if we need to (i.e., we've released the affinity). WDYT? Would remove the need for the calling code to do that.

Reply:
Yeah, I was thinking of doing something like that. Not sure why I ultimately chose to do it this way...

Reply:
Now I remember...The re-querying logic is no longer in this function as requested by https://github.com/projectcalico/libcalico-go/pull/999#discussion_r246602316. Should we bring the re-querying back in?

Reply:
I think that comment was in reference to something slightly different, but I'm happy to apply the same logic here for now. Let's carry on for now.